src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.Com.cs

```diff
@@ -103,15 +120,136 @@ public static IntPtr GetIUnknownForObject(object o)
         }
 
         [SupportedOSPlatform("windows")]
-        public static void GetNativeVariantForObject(object? obj, IntPtr pDstNativeVariant)
+        public static unsafe void GetNativeVariantForObject(object? obj, IntPtr pDstNativeVariant)
         {
-            throw new PlatformNotSupportedException(SR.PlatformNotSupported_ComInterop);
+            if (pDstNativeVariant == IntPtr.Zero)
+            {
+                throw new ArgumentNullException(nameof(pDstNativeVariant));
+            }
+
+            Variant* data = (Variant*)pDstNativeVariant;
+            if (obj == null)
+            {
+                data->VariantType = VarEnum.VT_EMPTY;
+                return;
+            }
+
+            if (obj is bool flag)
```

Comment:
This would look better with pattern matching:
```
switch (obj)
{
case bool value:
    data->AsBool = value;
    break;
case byte value;
    data->AsUi1 = value;
    break;
...
default:
...
}
```
Also, I would put int and string first since they are the most frequently used variants (the CoreCLR unmanaged impl has them first).